bootable-jar/runtime/src/main/java/org/wildfly/core/jar/runtime/InstallationCleaner.java

```diff
@@ -0,0 +1,196 @@
+                deleteDirectory();
+                throw e;
+            }
+        } else {
+            deleteDirectory();
+        }
+    }
+
+    private void deleteDirectory() throws IOException {
+        final Path installDir = environment.getJBossHome();
+        Files.walkFileTree(installDir, new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult visitFile(final Path file, final BasicFileAttributes attrs) {
+                try {
+                    // Don't delete the cleanup marker until we're ready to delete the directory
+                    if (!file.equals(cleanupMarker)) {
+                        Files.delete(file);
+                    }
+                } catch (IOException e) {
+                    logger.cantDelete(file.toString(), e);
+                }
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult postVisitDirectory(final Path dir, final IOException exc) {
+                try {
+                    if (dir.equals(installDir)) {
+                        // We have to delete the marker before we can delete the directory
+                        Files.deleteIfExists(cleanupMarker);
+                    }
+                    Files.delete(dir);
+                } catch (IOException e) {
+                    logger.cantDelete(dir.toString(), e);
+                }
+                return FileVisitResult.CONTINUE;
+            }
+        });
+    }
+
+    private void newProcess() throws IOException {
+        // Start a new process which will clean up the install directory. This is done in a new process in cases where
+        // this process may hold locks on to resources that need to be cleaned up.
+        final String[] cmd = {
+                getJavaCommand(),
+                "-cp",
+                // Use the current class path as it should just have the bootable JAR on it and this is where the
+                // CleanupProcess is located.
+                System.getProperty("java.class.path"),
+                "org.wildfly.core.jar.boot.CleanupProcessor",
+                environment.getJBossHome().toString(),
+                Integer.toString(retries)
+        };
+        final ProcessBuilder builder = new ProcessBuilder(cmd)
+                .redirectError(ProcessBuilder.Redirect.INHERIT)
+                .redirectOutput(ProcessBuilder.Redirect.INHERIT)
+                .directory(new File(System.getProperty("user.dir")));
+        builder.start();
+    }
+
+    private String getJavaCommand() {
+        final Path javaHome = Paths.get(System.getProperty("java.home"));
+        final Path java;
+        if (environment.isWindows()) {
+            java = javaHome.resolve("bin").resolve("java.exe");
+        } else {
+            java = javaHome.resolve("bin").resolve("java");
+        }
+        if (Files.exists(java)) {
+            return java.toString();
+        }
+        return "java";
```

Comment:
Should it be java.exe for Windows?

Reply:
I thought about that, but I know on Windows from CMD `java` works. That said `java.home` should never be a missing property and this is a fallback that should never happen. I should probably just verify the full path exists and then fall back though to be safe.